### PR TITLE
Add "Interrupt driven" mode to R-Car UART driver

### PR DESCRIPTION
**Pull request by aaillet**
This PR add the support of "Interrupt driven UART" to the R-Car UART driver and enable 
it in the related Kconfig.
 
The driver is supporting nearly all the methods that are described in the "Interrupt driven UART" 
part of the uart_driver_api.

This new version of the driver has been tested on H3ULCB board by running "uart_basic_api" test suite.

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

### CODE REVIEW

File: drivers/serial/uart_rcar.c [hunk 1]
```diff
@@ -216,7 +215,7 @@ static int uart_rcar_configure(const struct device *dev,
 		     SCFCR_MCE | SCFCR_TFRST | SCFCR_RFRST);
 	uart_rcar_write_16(config, SCFCR, reg_val);
 
-	/* Enable Transmit & Receive */
+	/* Enable Transmit & Receive + Interrupts disable */
```
Comment (julien-massot) [inline drivers/serial/uart_rcar.c]: Enable Transmit & Receive, disable Interrupts

File: boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7_defconfig [hunk 2]
```diff
@@ -13,11 +13,11 @@ CONFIG_GPIO_RCAR=y
...
 CONFIG_UART_CONSOLE=y
+CONFIG_UART_INTERRUPT_DRIVEN=y
 CONFIG_CAN=y
 CONFIG_CAN_RCAR=y
 CONFIG_PINMUX=y
 CONFIG_CAN=y
 CONFIG_CAN_RCAR=y
 CONFIG_I2C=y
 CONFIG_I2C_RCAR=y
-
```
Comment (julien-massot) [inline boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7_defconfig]: ?

File: drivers/serial/uart_rcar.c [hunk 3]
```diff
@@ -258,33 +265,256 @@ static int uart_rcar_init(const struct device *dev)
...
+
+	uint16_t reg_val;
+
+	reg_val = uart_rcar_read_16(config, SCSCR);
+	reg_val &= ~(SCSCR_REIE);
+	uart_rcar_write_16(config, SCSCR, reg_val);
+}
+
+static int uart_rcar_irq_is_pending(const struct device *dev)
+{
```
Comment (julien-massot) [inline drivers/serial/uart_rcar.c:414]: Maybe a comment may help there

File: drivers/serial/uart_rcar.c [hunk 4]
```diff
@@ -258,33 +265,256 @@ static int uart_rcar_init(const struct device *dev)
...
+	reg_val = uart_rcar_read_16(config, SCSCR);
+	reg_val &= ~(SCSCR_RIE);
+	uart_rcar_write_16(config, SCSCR, reg_val);
+}
+
+static int uart_rcar_irq_rx_ready(const struct device *dev)
+{
+	const struct uart_rcar_cfg *config = DEV_UART_CFG(dev);
+
+	return !!(uart_rcar_read_16(config, SCFSR) & SCFSR_RDF);
```
Comment (julien-massot) [inline drivers/serial/uart_rcar.c:388]: no sure we need parenthesis there

File: drivers/serial/uart_rcar.c (hunk 3 above)
Comment (aaillet) [reply drivers/serial/uart_rcar.c:414]: already commented in uart.h

File: drivers/serial/uart_rcar.c (hunk 4 above)
Comment (julien-massot) [reply drivers/serial/uart_rcar.c:388]: miss the & SCFSR_RDF
yes we need parenthesis

File: boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7_defconfig (hunk 2 above)
Comment (aaillet) [reply boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7_defconfig]: fixed thx

**Review by julien-massot — APPROVED** [review]

**Pull request merged**